### PR TITLE
Fix parsing error on urls without path.

### PR DESCRIPTION
**Pull request by jkotalik**
Fix for #405. At this point there are no tests for the Internal namespace, hence I didn't add any for now. We should consider adding them in the future?

### CODE REVIEW

**Comment by Tratcher** [comment]
Test? I'm curious to see the request that triggers this.

**Comment by jkotalik** [comment]
The issue is when a full url is provided into the request line. If the url doesn't include a slash after the hostname, it would fail to parse. Ex: `GET http://localhost:8080 HTTP/1.1` would throw an exception. @Tratcher I was thinking it may be correct to make the path "/" instead of empty?

**Comment by Tratcher** [comment]
Yes, make Path /

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/Listener/RequestTests.cs [hunk 1]
```diff
@@ -115,6 +115,21 @@ public class RequestTests
...
+        public async Task Request_FullUriInRequestLine_ParsesPath()
+        {
+            string root;
+            using (var server = Utilities.CreateHttpServerReturnRoot("/", out root))
+            {
+                // Send a HTTP request with the request line:
+                // GET http://localhost:5001 HTTP/1.1
+                var responseTask = SendSocketRequestAsync(root, root);
+                var context = await server.AcceptAsync(Utilities.DefaultTimeout);
+                Assert.Equal("/", context.Request.Path);
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/Listener/RequestTests.cs]: Also assert that PathBase is PathString.Empty.

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/Listener/RequestTests.cs [hunk 2]
```diff
@@ -115,6 +115,21 @@ public class RequestTests
         }
 
         [ConditionalFact]
+        public async Task Request_FullUriInRequestLine_ParsesPath()
+        {
+            string root;
+            using (var server = Utilities.CreateHttpServerReturnRoot("/", out root))
+            {
+                // Send a HTTP request with the request line:
+                // GET http://localhost:5001 HTTP/1.1
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/Listener/RequestTests.cs]: Assert `root` does not have a trailing slash. Utilities may be refactored later and unintentionally break this.

File: shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs [hunk 3]
```diff
@@ -39,7 +42,8 @@ public static ArraySegment<byte> GetPath(byte[] raw)
                     if (pathStartIndex == -1)
```
Comment (Tratcher) [inline shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs]: Do you see any tests verifying the comment about "http://server?query=value/1/2" being blocked by HttpSys?

File: shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs [hunk 4]
```diff
@@ -2,11 +2,14 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Text;
 
 namespace Microsoft.AspNetCore.HttpSys.Internal
 {
     internal static class RawUrlHelper
     {
+        private static byte[] _forwardSlashPath = Encoding.ASCII.GetBytes("/");
```
Comment (JunTaoLuo) [inline shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs]: private static readonly?

File: shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs (hunk 3 above)
Comment (jkotalik) [reply shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs]: There were no tests for full uris at all. I'll try to see what happens.

Comment (jkotalik) [reply shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs]: It returns a 400, which makes sense.

Comment (Tratcher) [reply shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/RawUrlHelper.cs]: Ok, please add one test.

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs [hunk 5]
```diff
@@ -259,6 +259,44 @@ public class RequestTests
...
+                var responseStatusCode = response.Substring(9); // Skip "HTTP/1.1 "
+                Assert.Equal("200", responseStatusCode);
+            }
+        }
+
+        [ConditionalFact]
+        public async Task Request_FullUriInRequestLineWithSlashesInQuery_BlockedByHttpSys()
+        {
+            string root;
+            using (var server = Utilities.CreateHttpServerReturnRoot("/", out root, httpContext =>
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs]: nit: out var

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs [hunk 6]
```diff
@@ -259,6 +259,44 @@ public class RequestTests
...
+            {
+                return Task.FromResult(0);
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs]: nit: Task.CompletedTask

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs [hunk 7]
```diff
@@ -259,6 +259,44 @@ public class RequestTests
...
+            }))
+            {
+                // Send a HTTP request with the request line:
+                // GET http://localhost:5001?query=value/1/2 HTTP/1.1
+                // Should return a 400 as it is a client error
+                var response = await SendSocketRequestAsync(root, root + "?query=value/1/2");
+                var responseStatusCode = response.Substring(9); // Skip "HTTP/1.1 "
+                Assert.Equal("400", responseStatusCode);
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs]: nit: use HttpStatusCodes (there's an enum somewhere that's named something like that)

**Comment by jkotalik** [comment]
Yeah those nits can be applied across the entire test suite. Either I have better style now or keep it consistent across the project. ¯\_(ツ)_/¯

**Comment by JunTaoLuo** [comment]
Slightly prefer if you start using the better style, but it doesn't really matter to me.

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs (hunk 7 above)
Comment (jkotalik) [reply test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestTests.cs]: https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/StatusCodes.cs

**Pull request merged**